### PR TITLE
Add CORS, and add CORS gtk settings

### PR DESCRIPTION
**Pull request by BrandonGillis**
This commit add the ability to enable/disable cross-origin request.
It also allow to add allowed origin within deluge gtk-ui.

image
#4

### CODE REVIEW

**Comment by idlesign** [comment]
Thank you. 
Almost ready to merge, but I have some remarks (mostly about style).

File: setup.py [hunk 1]
```diff
@@ -24,6 +24,8 @@
 __url__ = 'https://github.com/idlesign/deluge-webapi'
 
 
+__pkg_data__ = {__plugin_name__.lower(): ["data/*"]}
```
Comment (idlesign) [inline setup.py]: Style: here and elsewhere please use consistent string literals quoting.

What I use: 
```
'string literal'
'''multiline string literal'''
"""docstrings"""
```

File: webapi/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-VERSION = (0, 2, 1)
+VERSION = (0, 2, 2)
```
Comment (idlesign) [inline webapi/__init__.py]: That should rather be 0.3.0 semantically.

File: webapi/common.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+def get_resource(filename):
+    import pkg_resources, os
```
Comment (idlesign) [inline webapi/common.py]: No special need to import in function.

File: webapi/core.py [hunk 4]
```diff
@@ -1,20 +1,95 @@
...
 
 class Core(CorePluginBase):
 
     def enable(self):
         """"""
         LOGGER.info('Enabling WebAPI plugin CORE ...')
+        self.JSON_instance = component.get('JSON')
+        self.patched = False
+        self.config = deluge.configmanager.ConfigManager("webapi.conf", DEFAULT_PREFS)
+        if (self.config["enable_cors"] == True):
```
Comment (idlesign) [inline webapi/core.py]: `if self.config['enable_cors']:` would do. Here and elsewhere no need in explicit checks.

File: webapi/core.py [hunk 5]
```diff
@@ -1,20 +1,95 @@
...
+        self.config.save()
 
     def update(self):
         """"""
+
+    @export
+    def set_config(self, config):
+        "sets the config dictionary"
+        for key in config.keys():
+            self.config[key] = config[key]
```
Comment (idlesign) [inline webapi/core.py:41]: That seems to be the same as `dict.update(dict)`

File: webapi/core.py [hunk 6]
```diff
@@ -1,20 +1,95 @@
...
+        self.config.save()
+        if (self.config["enable_cors"] == True):
+            self.patch_web_ui()
+        elif (self.config["enable_cors"] == False):
```
Comment (idlesign) [inline webapi/core.py]: `elif not self.config['enable_cors']:` would do.

File: webapi/core.py [hunk 7]
```diff
@@ -1,20 +1,95 @@
...
+from deluge.ui.web.json_api import JSON
+from twisted.web import http, server
+import deluge.configmanager
+from deluge.core.rpcserver import export
 
 
 LOGGER = logging.getLogger(__name__)
 
+DEFAULT_PREFS = {
+    "enable_cors"         : False,
```
Comment (idlesign) [inline webapi/core.py]: In Python we usually do not do columns.

File: webapi/core.py [hunk 8]
```diff
@@ -1,20 +1,95 @@
...
+        self.old_send_request = self.JSON_instance._send_response
+        self.JSON_instance.render = self.render_patch
+        self.JSON_instance._send_response = self._send_response_patch
+        self.patched = True
+
+    def unpatch_web_ui(self):
+        if self.patched == False:
+            return
+        LOGGER.info('Unpatching webui for CORS...')
+        if (self.old_render):
```
Comment (idlesign) [inline webapi/core.py]: No need in brackets.

File: webapi/gtkui.py [hunk 9]
```diff
@@ -13,7 +19,96 @@ class GtkUI(GtkPluginBase):
...
+    def on_apply_prefs(self):
+        log.debug("applying prefs for WebAPI")
+        config = {}
+        if self.glade.get_widget("enable_cors").get_active():
+            config["enable_cors"] = True
+        else:
+            config["enable_cors"] = False
+
+        config["allowed_origin"] = []
+        if len(self.model) != 0:
```
Comment (idlesign) [inline webapi/gtkui.py]: Do we really need `len()` here?

File: webapi/gtkui.py [hunk 10]
```diff
@@ -13,7 +19,96 @@ class GtkUI(GtkPluginBase):
...
+            # remove all the entries in the model
+            for i in range(len(self.model)):
+                iter = self.model.get_iter(i)
```
Comment (idlesign) [inline webapi/gtkui.py]: Here and elsewhere `iter` is a built-in function, and we'd rather not to shadow it.

File: webapi/gtkui.py [hunk 11]
```diff
@@ -13,7 +19,96 @@ class GtkUI(GtkPluginBase):
...
+        config = {}
+        if self.glade.get_widget("enable_cors").get_active():
+            config["enable_cors"] = True
+        else:
+            config["enable_cors"] = False
+
+        config["allowed_origin"] = []
+        if len(self.model) != 0:
+            # remove all the entries in the model
+            for i in range(len(self.model)):
```
Comment (idlesign) [inline webapi/gtkui.py]: May `enumerate()` be enough?

File: webapi/gtkui.py [hunk 12]
```diff
@@ -13,7 +19,96 @@ class GtkUI(GtkPluginBase):
...
+                iter = self.model.get_iter(0)
+                self.model.remove(iter)
+        for domain in config["allowed_origin"]:
+            self.model.append([domain])
+
+    def create_listview(self):
+        self.model = gtk.ListStore(str)
+        self.list = self.glade.get_widget('treeview1')
```
Comment (idlesign) [inline webapi/gtkui.py]: Maybe something more verbose than `treeview1`?

File: webapi/webui.py [hunk 13]
```diff
@@ -5,11 +5,14 @@
 from deluge import component, common
 from deluge.ui.client import client
 from deluge.ui.web.json_api import export as export_api
+from deluge.ui.web.json_api import JSON
```
Comment (idlesign) [inline webapi/webui.py]: Are these new imports used in the module?

**Review by idlesign — CHANGES_REQUESTED** [review]

**Comment by BrandonGillis** [comment]
Thanks for your remarks, I'm gonna make the changes needed.

File: webapi/core.py (hunk 5 above)
Comment (BrandonGillis) [reply webapi/core.py:41]: This is not possible, as self.config is of type deluge.configmanager.ConfigManager and not directly a dict

**Review by idlesign — APPROVED** [review]

**Comment by idlesign** [comment]
Thank you. Merged.

**Pull request merged**